src/diffusers/models/transformers/sana_transformer_2d.py

```diff
@@ -0,0 +1,702 @@
+        else:
+            H, W = HW
+
+        hidden_states = hidden_states.reshape(B, H, W, C).permute(0, 3, 1, 2)
+
+        hidden_states = self.conv_inverted(hidden_states)
+        hidden_states = self.nonlinearity(hidden_states)
+
+        hidden_states = self.conv_depth(hidden_states)
+        hidden_states, gate = torch.chunk(hidden_states, 2, dim=1)
+        hidden_states = hidden_states * self.nonlinearity(gate)
+
+        hidden_states = self.conv_point(hidden_states)
+        hidden_states = hidden_states.reshape(B, C, N).permute(0, 2, 1)
+
+        return hidden_states
+
+
+# Modified from diffusers.models.attention.BasicTransformerBlock
+@maybe_allow_in_graph
+class SanaLinearTransformerBlock(nn.Module):
+    r"""
+    A Transformer block following the Linear Transformer architecture, introduced in Sana
+
+    Reference: https://arxiv.org/abs/2410.10629
+
+    Parameters:
+        dim (`int`): The number of channels in the input and output.
+        num_attention_heads (`int`): The number of heads to use for multi-head attention.
+        attention_head_dim (`int`): The number of channels in each head.
+        context_pre_only (`bool`): Boolean to determine if we should add some blocks associated with the
+            processing of `context` conditions.
+    """
+
+    def __init__(
+        self,
+        dim: int,
+        num_attention_heads: int,
+        attention_head_dim: int,
+        dropout=0.0,
+        num_cross_attention_heads: Optional[int] = None,
+        cross_attention_head_dim: Optional[int] = None,
+        cross_attention_dim: Optional[int] = None,
+        activation_fn: tuple = ("silu", "silu", None),
+        num_embeds_ada_norm: Optional[int] = None,
+        attention_bias: bool = False,
+        upcast_attention: bool = False,
+        norm_type: str = "ada_norm_single",
+        norm_elementwise_affine: bool = False,
+        norm_eps: float = 1e-6,
+        attention_out_bias: bool = True,
+        use_pe: bool = False,
+        num_positional_embeddings: Optional[int] = None,
+        expand_ratio: float = 2.5,
+    ):
+        super().__init__()
+        self.dim = dim
+        self.num_attention_heads = num_attention_heads
+        self.attention_head_dim = attention_head_dim
+        self.dropout = dropout
+        self.cross_attention_dim = cross_attention_dim
+        self.activation_fn = activation_fn
+        self.attention_bias = attention_bias
+        self.norm_elementwise_affine = norm_elementwise_affine
+
+        self.use_ada_layer_norm_single = norm_type == "ada_norm_single"
+        self.norm_type = norm_type
+        self.num_embeds_ada_norm = num_embeds_ada_norm
+
+        self.use_pe = use_pe
+        if use_pe:
+            self.pos_embed = SinusoidalPositionalEmbedding(dim, max_seq_length=num_positional_embeddings)
+        else:
+            self.pos_embed = None
+
+        self.norm1 = nn.LayerNorm(dim, elementwise_affine=False, eps=norm_eps)
+
+        self.attn1 = Attention(
```

Comment:
Should this be `MultiscaleLinearAttention` class that was introduced in DCAE PR, or the normal `Attention` class that is shipped in Diffusers for a long time now?

I see that the Sana attention processor is used here, so I think it should be former.

Reply:
It seems that the `MultiscaleLinearAttention` is more like to process a (B, C, H, W) tensor. Becasue, DC-AE is a Unet structure with most of the processors are Convs. Actually, in Sana, we prefer transformer format tensor, which is B, C, HW. The only difference with `Attention` is the attention processor. So I'm not sure which one is more suitable here. I will remain the `Attention` one here and let u decide which one is better.

Reply:
Okay, sounds good. I'll take a look soon and refactor